src/joypad.rs

```diff
@@ -33,20 +33,12 @@ impl Joypad {
     }
 
     fn map_buttons(buttons: [Key; 4], window: &Window) -> u8 {
-        let mut sum = 0;
-        if window.is_key_down(buttons[0]) {
-            sum += 1
-        }
-        if window.is_key_down(buttons[1]) {
-            sum += 2
-        }
-        if window.is_key_down(buttons[2]) {
-            sum += 4
-        }
-        if window.is_key_down(buttons[3]) {
-            sum += 8
-        }
-        !sum & 0x0F
+        !buttons
```

Comment:
Unfortunately iterators have a noticeable slowdown in debug builds so I've refactored most of the code to avoid using them. Other than that, the rest of the changes look good! Will merge once this change is reverted.

Reply:
I'm not home for a while, so in case you can pull this PR and revert the last commit.